sql/spanner/inspect.go

```diff
@@ -0,0 +1,771 @@
+		b.WriteByte('$')
+		b.WriteString(strconv.Itoa(start + i))
+	}
+	return b.String()
+}
+
+func defaultExpr(c *schema.Column, x string) schema.Expr {
+	switch {
+	case sqlx.IsLiteralBool(x), sqlx.IsLiteralNumber(x), sqlx.IsQuoted(x, '\''):
+		return &schema.Literal{V: x}
+	default:
+		// Try casting or fallback to raw expressions (e.g. column text[] has the default of '{}':text[]).
+		if v, ok := canConvert(c.Type, x); ok {
+			return &schema.Literal{V: v}
+		}
+		return &schema.RawExpr{X: x}
+	}
+}
+
+func canConvert(t *schema.ColumnType, x string) (string, bool) {
+	r := t.Raw
+	i := strings.Index(x, "::"+r)
+	if i == -1 || !sqlx.IsQuoted(x[:i], '\'') {
+		return "", false
+	}
+	q := x[0:i]
+	x = x[1 : i-1]
+	switch t.Type.(type) {
+	case *schema.BoolType:
+		if sqlx.IsLiteralBool(x) {
+			return x, true
+		}
+	case *schema.DecimalType, *schema.IntegerType, *schema.FloatType:
+		if sqlx.IsLiteralNumber(x) {
+			return x, true
+		}
+	case *schema.BinaryType, *schema.JSONType, *schema.SpatialType, *schema.StringType, *schema.TimeType, *StringType, *BytesType, *TimestampType:
+		return q, true
+	}
+	return "", false
+}
+
+type (
+	// ArrayType defines an array type.
+	// https://www.spannerql.org/docs/current/arrays.html
+	ArrayType struct {
+		schema.Type
+		T string
+	}
+
+	// A BytesType represents a BYTES type.
+	BytesType struct {
+		schema.Type
+		T         string
+		Size      int
+		SizeIsMax bool
+	}
+
+	// CheckColumns attribute hold the column named used by the CHECK constraints.
+	// This attribute is added on inspection for internal usage and has no meaning
+	// on migration.
+	CheckColumns struct {
+		schema.Attr
+		Columns []string
+	}
+
+	// IndexType represents an index type.
+	// https://www.spannerql.org/docs/current/indexes-types.html
+	IndexType struct {
+		schema.Attr
+		T string // BTREE, BRIN, HASH, GiST, SP-GiST, GIN.
+	}
+
+	// IndexPredicate describes a partial index predicate.
+	// https://www.spannerql.org/docs/current/catalog-pg-index.html
+	IndexPredicate struct {
+		schema.Attr
+		P string
+	}
+
+	// IndexColumnProperty describes an index column property.
+	// https://www.spannerql.org/docs/current/functions-info.html#FUNCTIONS-INFO-INDEX-COLUMN-PROPS
+	IndexColumnProperty struct {
+		schema.Attr
+		// NullsFirst defaults to true for DESC indexes.
+		NullsFirst bool
+		// NullsLast defaults to true for ASC indexes.
+		NullsLast bool
+	}
+
+	// IndexStorageParams describes index storage parameters add with the WITH clause.
+	// https://www.spannerql.org/docs/current/sql-createindex.html#SQL-CREATEINDEX-STORAGE-PARAMETERS
+	IndexStorageParams struct {
+		schema.Attr
+		// AutoSummarize defines the authsummarize storage parameter.
+		AutoSummarize bool
+		// PagesPerRange defines pages_per_range storage
+		// parameter for BRIN indexes. Defaults to 128.
+		PagesPerRange int64
+	}
+
+	// A StringType represents a STRING type.
+	StringType struct {
+		schema.Type
+		T         string
+		Size      int
+		SizeIsMax bool
+	}
```

Comment:
I suggested Fraser on Slack to add `Attrs` to the `schema.StringType` and attached this information there. I can help with this if you want.

Reply:
I'll ping for assist when I get there!

Reply:
Ok, took a stab [here](https://github.com/ariga/atlas/pull/1171/commits/f1ba496cb69a82ec6629a05701dfb38fc4f6c4f1). Let me know if that's not what you had in mind!

Reply:
Looks good, but please name is as `MaxSize` and remove the boolean flag. The Existence of this attribute is enough to indicate it is configured to MaxSize (`sqlx.Has`).

See another example for this here: https://github.com/ariga/atlas/pull/1153#discussion_r976255182